crates/biome_js_analyze/src/lint/nursery/use_exhaustive_switch_cases.rs

```diff
@@ -0,0 +1,314 @@
+use biome_analyze::{
+    FixKind, Rule, RuleDiagnostic, RuleDomain, context::RuleContext, declare_lint_rule,
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::{
+    AnyJsCallArgument, AnyJsExpression, AnyJsStatement, AnyJsSwitchClause, JsSwitchStatement, T,
+};
+use biome_js_type_info::{Literal, Type, TypeInner};
+use biome_rowan::{AstNode, AstNodeList, BatchMutationExt, TriviaPieceKind};
+
+use crate::JsRuleAction;
+use crate::services::typed::Typed;
+
+declare_lint_rule! {
+    /// Require switch-case statements to exhaustive.
+    ///
+    /// When working with union types or enums in TypeScript, it's common to want to write a switch
+    /// statement intended to contain a case for each possible type in the union or the enum.
+    /// However, if the union type or the enum changes, it's easy to forget to modify the cases to
+    /// account for any new types.
+    ///
+    /// This rule reports when a switch statement over a value typed as a union of literals or as an
+    /// enum is missing a case for any of those literal types and does not have a default clause.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```ts
+    /// type Day =
+    ///   | 'Monday'
+    ///   | 'Tuesday'
+    ///   | 'Wednesday'
+    ///   | 'Thursday'
+    ///   | 'Friday'
+    ///   | 'Saturday'
+    ///   | 'Sunday';
+    ///
+    /// declare const day: Day;
+    /// let result = 0;
+    ///
+    /// switch (day) {
+    ///   case 'Monday':
+    ///     result = 1;
+    ///     break;
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```ts
+    /// type Day =
+    ///   | 'Monday'
+    ///   | 'Tuesday'
+    ///   | 'Wednesday'
+    ///   | 'Thursday'
+    ///   | 'Friday'
+    ///   | 'Saturday'
+    ///   | 'Sunday';
+    ///
+    /// declare const day: Day;
+    /// let result = 0;
+    ///
+    /// switch (day) {
+    ///   case 'Monday':
+    ///     result = 1;
+    ///     break;
+    ///   case 'Tuesday':
+    ///     result = 2;
+    ///     break;
+    ///   case 'Wednesday':
+    ///     result = 3;
+    ///     break;
+    ///   case 'Thursday':
+    ///     result = 4;
+    ///     break;
+    ///   case 'Friday':
+    ///     result = 5;
+    ///     break;
+    ///   case 'Saturday':
+    ///     result = 6;
+    ///     break;
+    ///   case 'Sunday':
+    ///     result = 7;
+    ///     break;
+    /// }
+    /// ```
+    ///
+    pub UseExhaustiveSwitchCases {
+        version: "next",
+        name: "useExhaustiveSwitchCases",
+        language: "js",
+        recommended: false,
```

Comment:
I think this rule should be recommended, and it should emit a `Severity::Error` diagnostic. What do you think?

Reply:
I think we could follow noFloatingPromises, which also use the typing infrastructure and will be landed in v2.0. Should we promote this rule to a group then?

Reply:
Nevermind, `noFloatingPromises` is also recommended but not promoted from the nursery group yet. I changed `noExhaustiveSwitchCases` to recommended.